New Refresh Token Rotation content

tamigoodall opened the pull request:
Added new files
Added sections to /tokens/concepts/refresh-tokens.md

Review: 
- https://docs-content-staging-pr-8821.herokuapp.com/docs/tokens/concepts/refresh-tokens#refresh-token-rotation
- https://docs-content-staging-pr-8821.herokuapp.com/docs/tokens/concepts/refresh-token-rotation
- https://docs-content-staging-pr-8821.herokuapp.com/docs/tokens/guides/configure-refresh-token-rotation
- https://docs-content-staging-pr-8821.herokuapp.com/docs/tokens/guides/use-refresh-token-rotation
- https://docs-content-staging-pr-8821.herokuapp.com/docs/tokens/concepts/token-storage

<!---
Notes:
- Your PR should conform to our [Contributing Guidelines](https://github.com/auth0/docs/blob/master/CONTRIBUTING.md)
- If applicable, add details to the [update feed](https://github.com/auth0/docs/tree/master/updates)
- Make sure your PR gets deployed to a Heroku [Review App](https://github.com/auth0/docs/blob/master/CONTRIBUTING.md#review-apps) without errors
- Please include a short description
- If your PR is a work in progress, title it [DO NOT MERGE]
--->

tamigoodall committed:
Initial draft

Added new files
Added sections to /tokens/concepts/refresh-tokens.md

tamigoodall committed:
Edited text and descriptions

Edited text and descriptions

tamigoodall committed:
Update refresh-token-rotation.md

Moved link

tamigoodall committed:
Token Storage updates

Changed name/type of doc from Store Tokens (How-to) to Token Storage (Concept)
Updated links

tamigoodall committed:
Added sdk info

Added sdk info

TimFerrell commented on articles/tokens/concepts/refresh-token-rotation.md:
```diff
@@ -0,0 +1,45 @@
+
+Typically, ID and Access Tokens are obtained from the authorization server and cached in memory. Token renewal (due to evicted memory or expiration) is handled by the SDK, relieving the developer of this burden. Silent re-authentication is technically achieved by sending a `prompt=none` parameter on the authentication request and using a hidden iframe, provided that there is still an active user  ...
+
+Starting with ITP2, this became a problem with sessions being prematurely terminated in Safari. This is especially a concern if you want to offer long-lived sessions. To address this issue, you can use Refresh Token rotation to leverage the use of Refresh Tokens in the browser that adhere to the OAuth 2.0 BCP. 
```
This is also becoming an issue with other browsers too, such as Chrome.

stevehobbsdev replied:
The other one that people have reported to us through GitHub is Brave with it's Shield technology up (blocks third-party cookies by default).

stevehobbsdev replied:
> due to evicted memory or expiration

It might not be obvious what "evicted memory" really means here, might be worth adding "refreshing the browser" to this list.

tamigoodall replied:
Clarified to "memory cache eviction budgets" and added "refreshing the browsers"

TimFerrell commented on articles/tokens/guides/rotate-refresh-tokens-sdk.md:
```diff
@@ -0,0 +1,43 @@
+
+To take advantage of the Refresh Token rotation functionality, the option `useRefreshTokens` has been added to `createAuth0Client`, which defaults to `false`.
+
+* With this option set to `true`, the `offline_access` scope is automatically requested when using `loginWithRedirect(), loginWithPopup()` and `getTokenSilently()`. Furthermore, when `getTokenSilently()` is invoked and the access token has expired, the SDK will attempt to renew the ID and Access Tokens by calling the `/token` endpoint using the `refresh_token` grant type along with the Refresh To ...
```
If this exchange fails, what do we do @stevehobbsdev ?

stevehobbsdev replied:
It depends on at what point it fails.

If the exchange effectively fails because `useRefreshTokens` is `true` but we don't have a refresh token in the cache, then we fall back to the iframe method and try that (which would in turn fail if third-party cookies are blocked).

If there is a failure actually calling the `/token` endpoint with a refresh token, we would bubble that up to the developer.

tamigoodall replied:
Added clarification

stevehobbsdev commented on articles/tokens/concepts/refresh-token-rotation.md:
```diff
@@ -0,0 +1,45 @@
+
+* Developers have the option to use either memory or local storage to store all tokens (ID, access, refresh tokens). The SDK defaults to memory, as the more security-conscious option.
+
+* Developers now also have a config option to use Refresh Tokens to retrieve new access tokens. With this option enabled, SPA JS will now request the offline_access scope automatically, and call the token endpoint directly with a refresh token (if available). The default  iframe method with prompt=none will not be used if Refresh Tokens are enabled.
```
Should we use backticks here for code elements? I noticed they were used elsewhere.

```suggestion
* Developers now also have a config option to use Refresh Tokens to retrieve new access tokens. With this option enabled, SPA JS will now request the `offline_access` scope automatically, and call the token endpoint directly with a refresh token (if available). The default iframe method with `prompt=none` will not be used if Refresh Tokens are enabled.
```

stevehobbsdev replied:
> The default  iframe method with prompt=none will not be used if Refresh Tokens are enabled.

It's mentioned on another comment but we now fall back to using the iframe method if you have set `useRefreshTokens` to `true` but no refresh token is available in the cache. This is to try and help users to silently migrate to using refresh tokens without making them log in again.

tamigoodall replied:
I made the code formatting changes manually while making other updates. 
I also clarified the note about the iframe method.

stevehobbsdev commented on articles/tokens/concepts/refresh-token-rotation.md:
```diff
@@ -0,0 +1,45 @@
+
+* For refresh token support, the Auth0 tenant must have the Refresh Token Rotation feature [enabled](/tokens/guides/enable-refresh-token-rotation), and configured on a client using the Management API or Dashboard.
+
+* Developers have the option to use either memory or local storage to store all tokens (ID, access, refresh tokens). The SDK defaults to memory, as the more security-conscious option.
```
> as the more security-conscious option

Should we link to, or mention, anything about _why_ this is the more security-conscious option? Or is that made apparent in the rest of the doc?

tamigoodall replied:
I'll clean up that sentence to clarify.

tamigoodall committed:
Update token-storage.md

Fixed description

tamigoodall committed:
Incorporated reviewer changes

Incorporated reviewer changes

tamigoodall committed:
Final edits

Added configure-refresh-token-rotation
Added and updated links
Added sidebar hidden entries
Updated token-storage with more scenarios

tamigoodall committed:
Update in-memory-token-storage.png

Shrunk overly large diagram

tamigoodall committed:
Added attribute details

Added attribute details

esarafianou commented on articles/tokens/concepts/token-storage.md:
```diff
@@ -0,0 +1,118 @@
+
+### Browser in-memory scenarios
+
+Auth0 recommends that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) when storing tokens in memory to emulate private methods. Use [auth0-sp-js](https://github.com/auth0/auth0-spa-js). This is the most secure method for browser storage, however it **does not** provide persistence across page refreshes and bro ...
```
```suggestion
Auth0 recommends that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) when storing tokens in memory to emulate private methods. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js). This is the most secure method for browser storage, however it **does not** provide persistence across page refreshes and browser tabs. 
```

esarafianou commented on articles/tokens/concepts/token-storage.md:
```diff
@@ -0,0 +1,118 @@
+::: panel Reduce Security Risks with Local Storage
+If the SPA is using Implicit (although it is recommended that authorization code with PKCE is used) or Hybrid Flows, the absolute token expiration time can be reduced. This will reduce the impact of a reflected XSS (but not of a persistent one). To do so, go to **Dashboard > APIs > Settings > Token Expiration For Browser Flows (Seconds)**.
+
+Reduce the amount of third party JavaScript code included from a source outside your domain to the minimum needed (such as links to jQuery, Bootstrap, Google Analytics etc.) Reducing third-party JS code reduces the possibility of an XSS vulnerability. Performing Subresource Integrity checking (SRI) in third-party scripts (where possible) to verify that the resources fetched are delivered without  ...
```
Let's add a link in Subresource Integrity checking (SRI): https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity

tamigoodall replied:
Added

tamigoodall committed:
Apply suggestions from code review

Eva's typo fix

Co-Authored-By: Eva Sarafianou <eva.sarafianou@gmail.com>

tamigoodall committed:
Split file into two files

Split file into new file - use-refresh-token-rotation
Added link to SRI
Reordered section on refresh-tokens
Fixed frontmatter on use-refresh-tokens

tamigoodall committed:
Merge branch 'refresh-token-rotation' of https://github.com/auth0/docs into refresh-token-rotation

tamigoodall committed:
Update token-storage.md

Merged and/or removed redundant content

tamigoodall committed:
Shifted focus from SPA-specific descriptions

Shifted focus from SPA and/or called it out specifically

tamigoodall committed:
Added note about revoking RTs

Added note about revoking RTs

tamigoodall committed:
Added new file

Disable Refresh Token Rotation

tamigoodall committed:
Fixed typo in link

Fixed typo in link

tamigoodall committed:
Update token-storage.md

Cleaned up first scenario example

tamigoodall committed:
Update sidebar.yml

Added new file entries

tamigoodall committed:
Update refresh-token-rotation.md

Removed unnecessary h2s

esarafianou commented on articles/tokens/concepts/token-storage.md:
```diff
@@ -0,0 +1,100 @@
+
+### Browser in-memory scenarios
+
+Auth0 recommends that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) when storing tokens in memory to emulate private methods. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js). 
```
```suggestion
Auth0 recommends storing tokens in browser memory as the most secure option. Using [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) to handle the transmission and storage of tokens is the best way to protect the tokens, as Web Workers run in a separate global scope than the rest of the application. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js) whose default storage option is in-memory storage leveraging Web Workers.

If you cannot use Web Workers, Auth0 recommends as an alternative that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) to emulate private methods.
```

dctoon commented on articles/tokens/concepts/refresh-token-rotation.md:
```diff
@@ -0,0 +1,44 @@
+* [Resource Owner Password Credentials Exchange](/api-auth/tutorials/adoption/password)
+* [Device Authorization Flow](/flows/concepts/device-auth)
+
+Typically, ID and Access Tokens are obtained from the authorization server and cached in memory. Token renewal (due to refreshing the browser, memory cache eviction budgets, or expiration) is handled by the SDK. Silent re-authentication is technically achieved by sending a `prompt=none` parameter upon the authentication request and using a hidden iframe, provided that there is an active user sess ...
```
I'm not sure if this makes sense here, because this is specific to SPA's. A traditional and native app don't use the _hidden iframe_ concept. A note could be added that the recommendation for SPA's is to use the _PKCE_ with rotating Refresh Tokens.

Additionally I would callout a section around _intelligent tracking prevention_, where cookies could be _lost_ due to the ITP browser behaviour. Because of that the Auth0 session is lost and the user has to re-authenticate, where with a rotating Refresh Token there is no link with the Auth0 session, and therefore isn't affected by ITP mechanisms.

tamigoodall replied:
I moved some text around and added the ITP information.

dctoon commented on articles/tokens/concepts/refresh-tokens.md:
```diff
@@ -79,8 +69,31 @@ For information on using Refresh Tokens with our mobile SDKs, see:
+  - [Resource Owner Password Grant](/api-auth/grant/password)
+  - [Device Authorization Flow](/flows/concepts/device-auth)
+
+* A SPA (normally implementing [Implicit Flow](/flows/concepts/implicit)) should not ever receive a Refresh Token. A Refresh Token is essentially a user credential that allows a user to remain authenticated indefinitely. This sensitive information should be stored securely and *not* exposed client-side in a browser.
```
> A SPA (normally implementing [Implicit Flow](/flows/concepts/implicit)) should not ever receive a Refresh Token.

When I don't have a sense of what the _implicit flow_ is, I would interpret that Refresh Tokens don't work for SPA's, but they do when using the _code_ flow. Also the [BCP](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-13#section-3.1.2) recommends using the _code_ flow, so maybe we can clarify that.

tamigoodall replied:
I removed that bullet.

dctoon commented on articles/tokens/guides/revoke-refresh-tokens.md:
```diff
@@ -122,6 +122,14 @@ DELETE https://${account.namespace}/api/v2/device-credentials/{id}
+::: note
+When using Refresh Token rotation, If a previously invalidated token is used, the entire set of Refresh Tokens issued since that invalidated token was issued will immediately be revoked, requiring the end-user to re-authenticate.
+
+- Use the `/oauth/revoke` endpoint to revoke a Refresh Token. 
```
Also revokes the entire grant.

dctoon commented on articles/tokens/guides/configure-refresh-token-rotation.md:
```diff
@@ -0,0 +1,79 @@
+
+If a previously invalidated token is used, the entire set of Refresh Tokens issued since that invalidated token was issued will immediately be revoked, requiring the end-user to re-authenticate.
+
+- Use the `/oauth/revoke` endpoint to revoke a Refresh Token. 
```
Also revokes the entire grant.

dctoon commented on articles/tokens/guides/revoke-refresh-tokens.md:
```diff
@@ -122,6 +122,14 @@ DELETE https://${account.namespace}/api/v2/device-credentials/{id}
+
+- Use the `/oauth/revoke` endpoint to revoke a Refresh Token. 
+
+- Use the `/api/v2/device-credentials` endpoint to revoke Refresh Tokens configured for rotation. This endpoint revokes the entire grant not just a specific token.
```
Any reason to callout the `/api/v2/device-credentials`? Same comment applies for the section _Use Refresh Token Rotation_.

dctoon commented on articles/tokens/concepts/refresh-token-rotation.md:
```diff
@@ -0,0 +1,44 @@
+---
```
Doesn't it make sense add a new PKCE section on the page https://auth0.com/docs/architecture-scenarios/spa-api/part-1#implicit-grant, detailing the flow with a rotating Refresh Token?

tamigoodall replied:
It does make sense. Can you provide that detail @dctoon ?

tamigoodall replied:
Got it via Slack - thanks Toon!

tamigoodall committed:
Update articles/tokens/concepts/token-storage.md

Eva's update

Co-Authored-By: Eva Sarafianou <eva.sarafianou@gmail.com>

tamigoodall committed:
Toon's comments incorporated

Toon's review comments incorporated

esarafianou commented on articles/tokens/concepts/token-storage.md:
```diff
@@ -0,0 +1,102 @@
+
+### Browser local storage scenarios
+
+Storing tokens in browser local storage provides persistence across page refreshes and browser tabs, however if an attacker can achieve running JavaScript in the SPA, they can retrieve the tokens stored in local storage using a cross-site scripting (XSS) attack. A vulnerability leading to a successful XSS attack can be either in the SPA source code or in any third-party JavaScript code (such as b ...
```
```suggestion
Storing tokens in browser local storage provides persistence across page refreshes and browser tabs, however if an attacker can achieve running JavaScript in the SPA using a cross-site scripting (XSS) attack, they can retrieve the tokens stored in local storage. A vulnerability leading to a successful XSS attack can be either in the SPA source code or in any third-party JavaScript code (such as bootstrap, jQuery, or Google Analytics) included in the SPA.
```

tamigoodall replied:
I added this manually - as I had done it already based on your Slack comment.

tamigoodall committed:
Minor changes

Minor changes

tamigoodall committed:
Update token-storage.md

Added warning about XSS attacks with browser local storage option

tamigoodall committed:
Update token-storage.md

Moved XSS attack info

tamigoodall committed:
Update token-storage.md

Moved warning

dctoon commented on articles/tokens/guides/disable-refresh-token-rotation.md:
```diff
@@ -0,0 +1,31 @@
+PATCH /api/v2/clients/{client_id}
+{
+  "refresh_token": {
+    "rotation_type": "reusable"
```
```suggestion
    "rotation_type": "non-rotating",
```
Probably my bad :-)

dctoon commented on articles/tokens/guides/configure-refresh-token-rotation.md:
```diff
@@ -0,0 +1,71 @@
+
+    | Attribute | Description |
+    | -- | -- |
+    | `rotation_type` | Text string: "rotating" or "reusable" |
```
```suggestion
    | `rotation_type` | Text string: "rotating" or "non-rotating" |
```

dctoon reviewed (commented):
Quick feedback, probably my bad 🙃

tamigoodall committed:
Apply suggestions from code review

"reusable" to "non-rotating"

Co-Authored-By: Toon De Coninck <toon@auth0.com>

tamigoodall committed:
Added Refresh Token Rotation panel

Mobile + API Architecture Scenario
Flow - Auth0 Code PKCE

esarafianou commented on articles/tokens/concepts/token-storage.md:
```diff
@@ -0,0 +1,104 @@
+
+Auth0 recommends storing tokens in browser memory as the most secure option. Using [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) to handle the transmission and storage of tokens is the best way to protect the tokens, as Web Workers run in a separate global scope than the rest of the application. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js) whose defa ...
+
+If you cannot use Web Workers, Auth0 recommends as an alternative that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) to emulate private methods.
```
```suggestion
If you cannot use Web Workers, Auth0 recommends as an alternative that you use [JavaScript closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures) to emulate private methods.

 Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js) whose default storage option is in-memory storage. It leverages both Web Workers and JavaScript closures depending on the type of token.
```

stevehobbsdev replied:
> whose default storage option is in-memory storage leveraging Web Workers.

Just a question - the SPA SDK will leverage web workers where possible, except for IE11 where they are supported, but have other complications. Do we want to make reference to anything about that here, to the effect of "It leverages both Web Workers (in modern browsers), and..."

We could link to some readme docs that explains web workers and browser support specific to the SDK, in more detail.

esarafianou replied:
It's not only IE11. Access Tokens can't be stored in Web Workers. This document is about tokens in general.
Linking to some readme docs is a good idea!

stevehobbsdev replied:
The document is, but this paragraph in particular is in the context of the SPA SDK. Perhaps it's not the right place for that kind of detail, but let's link to specifics.

esarafianou replied:
Right and in the SPA SDK we use Web Workers for RT in modern browsers and JavaScript closures for ATs and for RTs in IE11. Am I missing something?

stevehobbsdev replied:
No you're correct, but the paragraph doesn't mention modern browsers - my question (and it was only a question) was whether or not it should.

esarafianou commented on articles/tokens/concepts/token-storage.md:
```diff
@@ -0,0 +1,104 @@
+
+### Browser in-memory scenarios
+
+Auth0 recommends storing tokens in browser memory as the most secure option. Using [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) to handle the transmission and storage of tokens is the best way to protect the tokens, as Web Workers run in a separate global scope than the rest of the application. Use [auth0-spa-js](https://github.com/auth0/auth0-spa-js) whose defa ...
```
```suggestion
Auth0 recommends storing tokens in browser memory as the most secure option. Using [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) to handle the transmission and storage of tokens is the best way to protect the tokens, as Web Workers run in a separate global scope than the rest of the application. 
```

tamigoodall replied:
Comitted

stevehobbsdev commented on articles/tokens/concepts/refresh-token-rotation.md:
```diff
@@ -0,0 +1,36 @@
+
+## SDK support
+
+For SPAs, use the [Auth0 SPA SDK](/libraries/auth0-spa-js) to perform silent re-authentication. Auth0 recommends that you use the [Authorization Code Flow with Proof Key for Code Exchange (PKCE)](/flows/concepts/auth-code-pkce).
```
> Auth0 recommends that you use the [Authorization Code Flow with Proof Key for Code Exchange (PKCE)](/flows/concepts/auth-code-pkce).

Is this here as a recommendation in general for SPAs, or related to the use of the SDK? If it's the former. I wonder if this sentence would be better of as a separate paragraph. When using the Auth0 SPA SDK the developer has no choice but to use Auth Code + PKCE, so to word it as a "recommendation" here feels redundant.

dctoon replied:
> Is this here as a recommendation in general for SPAs

☝️

stevehobbsdev commented on articles/tokens/concepts/refresh-token-rotation.md:
```diff
@@ -0,0 +1,36 @@
+
+With SPAs, you have the option to use either memory or local storage to store all tokens. The SDK defaults to memory for token storage. See [Token Storage](/tokens/concepts/token-storage) for details on which option will work for your app. 
+
+For iOS, Android, Windows and any other type of app, there are no changes. 
```
Do we mean here that that there are no changes for the developer to care about, or that there have been no changes to the SDK to support this?

If it's the latter, the iOS and Android SDKs had updates to support RTR, and new versions have been published.

stevehobbsdev commented on articles/tokens/concepts/refresh-tokens.md:
```diff
@@ -26,41 +26,29 @@ The Refresh Token behavior is applicable to [OIDC-conformant applications](/api-
 - The `parseHash` method, parses a URL hash fragment to extract the result of an Auth0 authentication response.
-- The `checkSession` method, attempts to get a new token from Auth0, using [silent authentication](/api-auth/tutorials/silent-authentication). For more details refer to [Using checkSession to Acquire New Tokens](/libraries/auth0js#using-checksession-to-acquire-new-tokens). An example is as follows:
+- The `checkSession` method, attempts to get a new token from Auth0, using [silent authentication](/api-auth/tutorials/silent-authentication). For more details refer to [Using checkSession to Acquire New Tokens](/libraries/auth0js#using-checksession-to-acquire-new-tokens). 
+- The Auth0-spa-js SDK supports Refresh Token rotation capabilities.
```
Suggestion, for consistency:

```suggestion
- The [Auth0 SPA SDK](/libraries/auth0-spa-js) supports Refresh Token rotation capabilities.
```

tamigoodall replied:
Commited

stevehobbsdev commented on articles/tokens/concepts/refresh-tokens.md:
```diff
@@ -26,41 +26,29 @@ The Refresh Token behavior is applicable to [OIDC-conformant applications](/api-
 
 ### Single-page apps
 
 [Silent Authentication](/api-auth/tutorials/silent-authentication) is the method that is used to refresh a token for web apps that execute in a browser. [Auth0.js](/libraries/auth0js), our client-side library, provides methods for this functionality:
```
> Auth0.js, our client-side library

I know this wasn't something that was changed, but should we alter the wording here to posit that this SDK is older and has effectively been superseded? I don't think these are the right words, but something to the effect of "Auth0.js, our older JS SDK.."

stevehobbsdev commented on articles/tokens/concepts/token-storage.md:
```diff
@@ -0,0 +1,104 @@
+
+- If the SPA backend cannot handle the API calls, the tokens should be stored in the SPA backend but the SPA needs to fetch the tokens from the backend to perform requests to the API. A protocol needs to be established between the backend and the SPA to allow the secure transfer of the token from the backend to the SPA.
+
+- If you have a SPA with **no** corresponding backend server, your SPA should request new tokens on login and store them in memory without any persistence. To make API calls, your SPA would then use the in-memory copy of the token.
```
> If you have a SPA

Throughout the docs there's a mix of "a SPA" and "an SPA" - I recommend having a sweep through and picking one style or the other, for consistency.

tamigoodall replied:
Good catch. I'll do that.

stevehobbsdev commented on articles/tokens/guides/configure-refresh-token-rotation.md:
````diff
@@ -0,0 +1,71 @@
+    npm install @auth0/auth0-spa-js
+    ```
+
+2. Use the Managment API to enable Refresh Token rotation:
````
Will this be required? Presumably by the time these docs hit live, the changes to Manhatten will be live and the admin will be able to enable this feature using the UI rather than using the management API.

randynasson replied:
Exactly. I also shared some screenshots of what the Admin controls look like and suggest that this be updated.

stevehobbsdev commented on articles/tokens/guides/configure-refresh-token-rotation.md:
````diff
@@ -0,0 +1,71 @@
+    ```text
+    npm install @auth0/auth0-spa-js
+    ```
+
````
We'll need another step in this sequence. Simply installing the SDK is not enough to start using refresh tokens, as the developer will need to enable the feature on the SDK for it to start sending the `offline_access` scope.

Here's what it looks like for the developer:

```js
const auth0 = await createAuth0Client({
  domain: '<YOUR AUTH0 DOMAIN>',
  client_id: '<YOUR CLIENT ID>',
  audience: '<YOUR API IDENTIFIER>',
  useRefreshTokens: true
});
```

Feel free to use whatever placeholders make sense. The key part is the `useRefreshTokens` parameter is set to `true`.

stevehobbsdev commented on articles/tokens/guides/disable-refresh-token-rotation.md:
```diff
@@ -0,0 +1,31 @@
+---
+# Disable Refresh Token Rotation
+
+Disable Refresh Token Rotation for each application using the Management API: 
```
Same question as above - will the admin be able to disable this feature through the dashboard instead of using the API?

randynasson replied:
Yes. Needs updated info.

stevehobbsdev commented on articles/tokens/guides/use-refresh-token-rotation.md:
```diff
@@ -0,0 +1,63 @@
+Silent re-authentication is achieved by sending a `prompt=none` parameter upon the authentication request and using a hidden iframe, provided that there is an active user session on the authorization server. The SDK uses the iframe method if you have set `useRefreshTokens` to `true` but no Refresh Token is available in the cache. This helps users to silently migrate to using Refresh Tokens withou ...
+
+::: note
+If the exchange fails because `useRefreshTokens` is `true` but there isn't a Refresh Token in the cache, then it falls back to the iframe method (which could also fail if third-party cookies are blocked).
```
Perfect!

randynasson committed:
Update refresh-token-rotation.md

randynasson committed:
Update refresh-token-rotation.md

tamigoodall committed:
Apply suggestions from code review

Reviewer changes

Co-Authored-By: Eva Sarafianou <eva.sarafianou@gmail.com>

tamigoodall committed:
Randy's Changes

Added state diagram place holder png
Updated silent authentication

tamigoodall committed:
Merge branch 'master' into refresh-token-rotation

tamigoodall committed:
Update customize-mfa-universal-login.md

Changed title of linked doc

tamigoodall committed:
Reuse detection description

Reuse detection description and diagram
SPA support description updates
Added detail from blog draft

tamigoodall committed:
Added new screenshots and steps

Updated configure-refresh-token-rotation to include steps for new Dashboard settings.
Added second step for Management API steps

tamigoodall committed:
Added note

_rtr_enabled include statement
Added to Token Storage and Use Refresh Token Rotation

tamigoodall committed:
Fixed includes nesting

Fixed includes nesting

tamigoodall committed:
Updated diagram

Updated diagram and steps

tamigoodall committed:
Added RTR recommend include

Added RTR recommend include

tamigoodall committed:
Added diagram

Added diagram
Rewrote intro for rtr

tamigoodall committed:
Update refresh-token-rotation.md

Added paragraph on long-lived RTs

tamigoodall committed:
Update refresh-tokens.md

Remove outdated limitation

tamigoodall committed:
Added links to new content

Added links to new content

tamigoodall committed:
Update silent-authentication.md

redirect fix

tamigoodall committed:
Typo fix

Typo fix

tamigoodall committed:
Minor edits

Updated Access Token Lifetime to add note
Fixed "Click Save Changes"

tamigoodall committed:
Update configure-refresh-token-rotation.md

Minor edits

tamigoodall committed:
Update configure-refresh-token-rotation.md

Fixed confusing wording

tamigoodall committed:
Update token-storage.md

Fixed error in link

tamigoodall committed:
Update token-storage.md

Fixed links

erjohnson commented on articles/tokens/concepts/refresh-token-rotation.md:
```diff
@@ -0,0 +1,74 @@
+
+## Maintaining user sessions in SPAs
+
+Until very recently, SPAs overcame the challenge of maintaining the user’s session by using the [Authorization Code Flow with PKCE](/flows/concepts/auth-code-pkce) in conjunction with [Silent Authentication](/api-auth/tutorials/silent-authentication). Recent developments in browser privacy technology, such as Intelligent Tracking Prevention (ITP) prevent access to the Auth0 session cookie, thereb ...
```
Maybe: SPAs maintained the user’s session by using the [Authorization Code Flow with PKCE](/flows/concepts/auth-code-pkce) along with [Silent Authentication](/api-auth/tutorials/silent-authentication).

erjohnson commented on articles/tokens/concepts/refresh-token-rotation.md:
This means you can safely...

erjohnson commented on articles/tokens/concepts/refresh-token-rotation.md:
This safeguards your app from replay attacks...

erjohnson commented on articles/tokens/concepts/refresh-token-rotation.md:
Maybe something like:

So it’s important that when a previously-used Refresh Token (already invalidated) is sent to the authorization server that the most recently issued Refresh Token is immediately invalidated as well. This prevents any Refresh Tokens in the same token family (all Refresh Tokens descending from the original Refresh Token issued for the client) from being used to get new Access Tokens.

This is a tough section since there's a lot of info to convey.

tamigoodall replied:
> So it’s important that the most recently issued Refresh Token is also immediately invalidated when a previously-used Refresh Token (already invalidated) is sent to the authorization server.

erjohnson commented on articles/tokens/concepts/refresh-tokens.md:
```diff
@@ -11,11 +11,11 @@ useCase:
 Auth0 issues an <dfn data-key="access-token">[Access Token](/tokens/concepts/access-tokens)</dfn> or an [ID Token](/tokens/concepts/id-tokens) in response to an [authentication request](/api-auth). You can use Access Tokens to make authenticated calls to a secured API, while the ID Token contains user profile attributes represented in the form of *claims*. Both are [JWTs](/tokens/concepts/jwts) a ...
 
-A Refresh Token is a special kind of token that can be used to obtain a renewed access token. You are able to request new access tokens until the Refresh Token is blacklisted. It’s important that refresh tokens are [stored securely](/tokens/guides/store-tokens) by the application because they essentially allow a user to remain authenticated forever.
+A Refresh Token is a special kind of token that can be used to obtain a renewed Access Token. You are able to request new access tokens until the Refresh Token is blacklisted. It’s important that refresh tokens are [stored securely](/tokens/concepts/token-storage) by the application because they essentially allow a user to remain authenticated forever.
```
A Refresh Token is a special kind of token used to obtain a renewed Access Token. You can request new access tokens until the Refresh Token is blacklisted. Applications must [store refresh tokens securely](/tokens/concepts/token-storage) because they essentially allow a user to remain authenticated forever.

erjohnson commented on articles/tokens/concepts/refresh-token-rotation.md:
When reuse is detected... -> Auth0 captures detected reuse events (`ferrt` indicating a failed exchange) in logs.

tamigoodall committed:
Eric's comments

Eric's comments
Sidebar

tamigoodall committed:
Update sidebar.yml

added children
